### PR TITLE
Warn about missing `default-language`

### LINKED ISSUE #9620: Warn on missing default-language
`default-language` has been optional for some time (see #6288). However, now that GHC has started introducing language editions like `GHC2021` and `GHC2024`, packages that do not specify a `default-language` ~~may be broken when GHC's default changes~~ will fall back to `Haskell98`, potentially unexpectedly. Thus it would be better if `cabal check` issued a warning about such packages.

See discussion from https://github.com/ghc-proposals/ghc-proposals/pull/613#issuecomment-1879188259 onwards. As @gbaz notes:
> We have a pretty clear plan of action that we can all agree on -- a pr to the cabal repo that creates a warning for missing language of at least "suspicious" severity -- which will lead to it being classified by `isHackageDistError` as something hackage would reject, and will lead to hackage, when it moves to using the new version of the cabal library, indeed rejecting the package.

### PR DESCRIPTION
**Pull request by ffaf1**
Closes #9620 

To accomodate GHC language editions (GHC2021, GHC2024, etc.).

---

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.
* [x] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.
* [x] Tests have been added. (*Ask for help if you don’t know how to write them! Ask for an exemption if tests are too complex for too little coverage!*)

---

**QA notes**

- Start with a `.cabal` file with `cabal-version: 3.4` and no `default-language` field.
- Old `cabal check` *will not* complain about missing default language.
- New `cabal check` *will* emit a `no-default-language` warning.

### CODE REVIEW

File: changelog.d/pr-9621 [hunk 1]
```diff
@@ -0,0 +1,9 @@
+synopsis: Warn about missing `default-language`
+packages: Cabal cabal-install
+prs: #9621
+issues: #9620
+description: {
+
+- `cabal check` will warn about missing `default-language` and not assume `Haskell 2010`.
```
Comment (mpickering) [inline changelog.d/pr-9621:7]: I am not sure this comment is accurate, if you see `Distribution.Simple.GHC.Internal`, not specifying a `default-language` defaults to `Haskell98`.

**Review by mpickering — CHANGES_REQUESTED** [review]

File: changelog.d/pr-9621 (hunk 1 above)
Comment (fgaz) [reply changelog.d/pr-9621:7]: #6926 says otherwise. maybe buildinfo gets populated anyway? we should test with -v3`

Comment (ffaf1) [reply changelog.d/pr-9621:7]: This is specifically about `check` though.

Would removing “ and not assume `Haskell 2010`” be ok for you?

File: Cabal/src/Distribution/PackageDescription/Check/Target.hs [hunk 2]
```diff
@@ -485,7 +485,6 @@ checkBuildInfoFeatures bi sv = do
   -- CheckSpecVer sv.
   checkP
     ( sv >= CabalSpecV1_10
-        && sv < CabalSpecV3_4
         && isNothing (defaultLanguage bi)
     )
     (PackageBuildWarning CVDefaultLanguageComponent)
```
[the remark this answers is not in the document]
Comment (ffaf1) [reply Cabal/src/Distribution/PackageDescription/Check/Target.hs:490]: It does

https://github.com/haskell/hackage-server/blob/1daf74881c2ab7820bd5be8008274490bd9ccd87/src/Distribution/Server/Packages/Unpack.hs#L297-L305

(and we should remember to change this bit to `isHackageDistError` as soon as Cabal 3.12 is released)

File: changelog.d/pr-9621 (hunk 1 above)
Comment (fendor) [reply changelog.d/pr-9621:7]: The Cabal changelog also claims it uses the compiler's default language: https://cabal.readthedocs.io/en/stable/file-format-changelog.html#cabal-version-3-4

File: Cabal/src/Distribution/PackageDescription/Check/Target.hs [hunk 3]
```diff
@@ -485,7 +485,6 @@ checkBuildInfoFeatures bi sv = do
   -- CheckSpecVer sv.
   checkP
     ( sv >= CabalSpecV1_10
-        && sv < CabalSpecV3_4
```
[the remark this answers is not in the document]
Comment (ffaf1) [reply Cabal/src/Distribution/PackageDescription/Check/Target.hs:488]: I am checking #6926 and I am just seeing a `^^^ availableSince CabalSpecV1_10 Nothing` w/r/t spec (specifically: there is nothing regarding CabalSpecV3_4).

More in general, I don’t think we are messing with the spec this time (compare with PackageInfo fiasco), we are just dealing with `check` behaviour, but I might be wrong.

`cabal check` raison d'être is (also) to mirror future Hackage behaviour; if Hackage decides *not* to accept packages withouth `default-language` for all cabal spec versions, `cabal check` should follow that.

**Comment by ffaf1** [comment]
Someone (I believe @alt-romes) pointed out in dev-call that current behaviour (out of sync with docs, alas) defaults to `Haskell98`.

I am closing this and bringing the discussion in the original issue #9620

**Comment by alt-romes** [comment]
I've opened #9668 to document that behaviour.

**Pull request closed without merging**